src/lib.rs

```diff
@@ -0,0 +1,450 @@
+                self.statements = Some(new_map);
+            }
+        };
+    }
+
+    pub fn get_statement(&self, name: &str) -> Option<&StatementBody> {
+        match self.statements {
+            Some(ref statements) => {
+                statements.get(name)
+            },
+            None => None
+        }
+    }
+}
+
+
+// Utilities for creating Blockly data structures
+
+pub fn program_from_xml(xml: &str) -> Program {
+    let mut program = Program::new();
+
+    let package: Package = parser::parse(xml).expect("Failed to parse XML!");
+    let document: Document = package.as_document();
+
+    let xml_element = get_xml_element(document);
+
+    for child in xml_element.children().iter() {
+        match child {
+            &ChildOfElement::Element(el) => {
+                let element_name = el.name().local_part();
+                match element_name {
+                    "block" => {
+                        program.groups.push(make_statement_body(el));
+                    },
+                    _ => {}
+                }
+            },
+            _ => {}
+        }
+    }
+
+    program
+}
+
+fn get_next_block_element(block_el: Element) -> Option<Element> {
+    let next_el: Option<Element> = {
+        let mut found: Option<Element> = None;
+        for child in block_el.children().iter() {
+            match child {
+                &ChildOfElement::Element(el) => {
+                    match el.name().local_part() {
+                        "next" => {
+                            found = Some(el);
+                            break;
+                        },
+                        _ => {}
+                    }
+                },
+                _ => {}
+            }
+        }
+        found
+    };
+
+    match next_el {
+        Some(next_el) => {
+            for child in next_el.children().iter() {
+                match child {
+                    &ChildOfElement::Element(el) => {
+                        match el.name().local_part() {
+                            "block" => {
+                                return Some(el);
+                            },
+                            _ => {}
+                        }
+                    },
+                    _ => {}
+                }
+            }
+            None
+        },
+        None => None
+    }
+}
+
+fn make_statement_body(first_block: Element) -> StatementBody {
+    let mut body = StatementBody::new();
+
+    // Create each block, put them into the statement body
+    let mut block_el: Element;
+    block_el = first_block;
+    loop {
+        body.blocks.push(make_block(block_el));
+        let next_block = get_next_block_element(block_el);
+        match next_block {
+            Some(next_block_el) => {
+                block_el = next_block_el;
+            },
+            None => {
+                break;
+            }
+        }
+    }
+
+    body
+}
+
+fn make_block(block_el: Element) -> Block {
```

Comment:
I would replace `Block::new()` with this function, like `pub fn new(block_el: Element) -> Self`, which would allow you to skip creating a block with an empty string for the `id` and `block_type`.

Reply:
That's a good point! All of my `make_*` functions are essentially `new` methods, so I might as well convert them for simplicity. I think I'll convert them to private in that case, because I don't want to leak the `sxd_document` dependency to consumers of my library.